src/test/e2e/fixtures/auth.setup.js

```diff
@@ -0,0 +1,150 @@
+const lockDirPath = path.join(__dirname, '..', 'loginCounts.lock'); // Directory for locking
+
+// Tracks whether a login has occurred for a given role.
+const hasLoggedIn = {};
+
+// Atomic read-modify-write with directory locking
+async function updateCounts(testFilePath) {
+  const maxRetries = 10;
+  const retryDelay = 100; // Milliseconds
+
+  for (let attempt = 0; attempt < maxRetries; attempt++) {
+    try {
+      fs.mkdirSync(lockDirPath);
+
+      let loginCounts = {};
+      try {
+        if (fs.existsSync(countsFilePath)) {
+          loginCounts = JSON.parse(fs.readFileSync(countsFilePath, 'utf8'));
+        }
+      } catch (readError) {
+        console.error('Error reading or parsing loginCounts.json:', readError);
+      }
+
+      if (!loginCounts[testFilePath]) {
+        loginCounts[testFilePath] = 0;
+      }
+      loginCounts[testFilePath]++;
+
+      try {
+        fs.writeFileSync(countsFilePath, JSON.stringify(loginCounts, null, 2), 'utf8');
+      } catch (writeError) {
+        console.error('Error writing loginCounts.json:', writeError);
+        continue; // Retry
+      }
+
+      fs.rmdirSync(lockDirPath);
+      return; // Success
+
+    } catch (error) {
+      if (error.code === 'EEXIST') {
+        await new Promise(resolve => setTimeout(resolve, retryDelay));
+      } else {
+        console.error('Error acquiring or releasing lock:', error);
+        throw error;
+      }
+    }
+  }
+  console.error(`Failed to acquire lock after ${maxRetries} attempts.`);
+  throw new Error(`Failed to acquire lock after ${maxRetries} attempts.`);
+}
+
+const roleCredentials = {
+  admin: {
+    username: process.env.OAUTH_USER,
+    password: process.env.OAUTH_USER_PASSWORD
+  },
+  viewer: {
+    username: process.env.OAUTH_VIEWER_USER,
+    password: process.env.OAUTH_USER_PASSWORD
+  },
+  superAdmin: {
+    username: process.env.OAUTH_SUPER_USER,
+    password: process.env.OAUTH_USER_PASSWORD
+  },
+  noRole: { username: process.env.OAUTH_TEST_USER_NO_ROLE, password: process.env.OAUTH_USER_PASSWORD }
+};
+
+exports.test = base.extend({
+  adminPage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'admin', testInfo);
+    await use(page);
+  },
+  viewerPage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'viewer', testInfo);
+    await use(page);
+  },
+  superAdminPage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'superAdmin', testInfo);
+    await use(page);
+  },
+  noRolePage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'noRole', testInfo);
+    await use(page);
+  },
+  page: async ({ page }, use) => {
+    await use(page);
+  }
+});
+
+async function loginWithRole(page, role, testInfo) {
+  const loginPage = new LoginPage(page);
+  const credentials = roleCredentials[role];
+  const testFilePath = testInfo.file;
+  const loginKey = role;
+
+  if (!hasLoggedIn[loginKey]) {
```

Comment:
I presume this is only set after the login is successful. If it fails then I imagine it will retry and not his this line (double checking, probably so!)